controllers/serviceinstance_controller.go

```diff
@@ -120,13 +124,69 @@ func (r *ServiceInstanceReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 		return r.createInstance(ctx, smClient, serviceInstance)
 	}
 
+	// Handle instance share change if needed
+	if instanceShareChangeNeedsToBeHandled(serviceInstance) {
+		if err := r.handleInstanceSharingChange(ctx, serviceInstance, smClient); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{Requeue: true, RequeueAfter: r.Config.PollInterval}, nil
+	}
+
 	// Update
 	if serviceInstance.Status.Ready == metav1.ConditionTrue {
 		return r.updateInstance(ctx, smClient, serviceInstance)
 	}
+
 	return ctrl.Result{}, nil
 }
 
+func instanceShareChangeNeedsToBeHandled(serviceInstance *servicesv1.ServiceInstance) bool {
+	return servicesv1.ShouldHandleSharing(serviceInstance.Spec.Shared, serviceInstance.Status.Shared) &&
+		serviceInstance.Status.Ready == metav1.ConditionTrue &&
+		!isShareFailed(serviceInstance.GetConditions())
```

Comment:
why do you need to check if shared is failed? you didnt relate to transient/non transient error
you should check what kind of error, because if it failed since the plan is not shareable there is no point on trying to share it again

Reply:
We agreed that any error shared is a final error.
And I do not want to get into an infinite loop - if shared failed - we don't try again.